syntax = "proto2";

message Trace {
    repeated Increment increment = 1;
}

message Increment {
    required int64 time_stamp       = 1;

    oneof increment {
        Transaction  transaction    = 2;
        Create       create         = 3;
        Delete       delete         = 4;
        BufferUpdate buffer_update  = 5;
        VSyncEvent   vsync_event    = 6;
    }
}

message Transaction {
    repeated Change change      = 1;

    required bool   synchronous = 2;
    required bool   animation   = 3;
}

message Change {
    required uint32 id = 1;

    oneof Change {
        PositionChange              position                   = 2;
        SizeChange                  size                       = 3;
        AlphaChange                 alpha                      = 4;
        LayerChange                 layer                      = 5;
        CropChange                  crop                       = 6;
        FinalCropChange             final_crop                 = 7;
        MatrixChange                matrix                     = 8;
        OverrideScalingModeChange   override_scaling_mode      = 9;
        TransparentRegionHintChange transparent_region_hint    = 10;
        LayerStackChange            layer_stack                = 11;
        HiddenFlagChange            hidden_flag                = 12;
        OpaqueFlagChange            opaque_flag                = 13;
        SecureFlagChange            secure_flag                = 14;
        DeferredTransactionChange   deferred_transaction       = 15;
    }
}

message PositionChange {
    required float x = 1;
    required float y = 2;
}

message SizeChange {
    required uint32 w = 1;
    required uint32 h = 2;
}

message AlphaChange {
    required float alpha = 1;
}

message LayerChange {
    required uint32 layer = 1;
}

message CropChange {
    required Rectangle rectangle = 1;
}

message FinalCropChange {
    required Rectangle rectangle = 1;
}

message MatrixChange {
    required float dsdx = 1;
    required float dtdx = 2;
    required float dsdy = 3;
    required float dtdy = 4;
}

message OverrideScalingModeChange {
    required int32 override_scaling_mode = 1;
}

message TransparentRegionHintChange {
    repeated Rectangle region = 1;
}

message LayerStackChange {
    required uint32 layer_stack = 1;
}

message HiddenFlagChange {
    required bool hidden_flag = 1;
}

message OpaqueFlagChange {
    required bool opaque_flag = 1;
}

message SecureFlagChange {
    required bool secure_flag = 1;
}

message DeferredTransactionChange {
    required uint32 layer_id     = 1;
    required uint64 frame_number = 2;
}

message Rectangle {
    required int32 left   = 1;
    required int32 top    = 2;
    required int32 right  = 3;
    required int32 bottom = 4;
}

message Create {
    required uint32 id    = 1;
    required string name  = 2;
    required uint32 w     = 3;
    required uint32 h     = 4;
}

message Delete {
    required uint32 id = 1;
}

message BufferUpdate {
    required uint32 id            = 1;
    required uint32 w             = 2;
    required uint32 h             = 3;
    required uint64 frame_number  = 4;
}

message VSyncEvent {
    required int64 when = 1;
}
